### PR TITLE
feat(i18n): expand fallback system

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

- What does this change?
- Be short and concise. Bullet points can help!
- Before/after screenshots can help as well.
- Don't forget a changeset! `pnpm exec changeset`

### CODE REVIEW

File: packages/astro/test/fixtures/rewrite-trailing-slash-never/src/pages/foo.astro [hunk 1]
```diff
@@ -1,3 +1,3 @@
 ---
-return Astro.rewrite("/")
+return Astro.rewrite("/base")
```
Comment (ematipico) [inline packages/astro/test/fixtures/rewrite-trailing-slash-never/src/pages/foo.astro:2]: This was an incorrect rewrite. The bug uncovered this, and I changed it. The fixture contains a `base` which means doing `/` should result to a 404 instead.

File: .changeset/lazy-feet-join.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  i18n: {
+    locals: ["en", "fr"],
+    defaultLocale: "en",
+    routing: {
+      prefixDefaultLocale: true
+    },
+    fallback: {
+      fr: "en"
+    },
+    fallbackType: "rewrite"
```
Comment (matthewp) [inline .changeset/lazy-feet-join.md:25]: Should this option be under `routing` since it pertains to how the default routing middleware works? If a user sets `routing: 'manual' then this option doesn't do anything, right?

File: packages/astro/src/core/routing/rewrite.ts [hunk 3]
```diff
@@ -12,44 +12,65 @@ export type FindRouteToRewrite = {
...
+		newUrl = payload;
 	} else if (payload instanceof Request) {
-		finalUrl = new URL(payload.url);
+		newUrl = new URL(payload.url);
 	} else {
-		finalUrl = new URL(payload, new URL(request.url).origin);
+		newUrl = new URL(payload, new URL(request.url).origin);
+	}
+	let pathname = newUrl.pathname;
+	if (base !== '/' && newUrl.pathname.includes(base)) {
```
Comment (matthewp) [inline packages/astro/src/core/routing/rewrite.ts:43]: should this be `startsWith`?

**Comment by sarah11918** [comment]
Re: the patch changeset, is it possible that people tried to "work around" this bug, and may have to update their projects now accordingly?

e.g. Could someone have written `return Astro.rewrite("/")` and *expect* it to include the base, thinking that this is just how the feature worked, and now they need to add it in manually to preserve their existing behaviour? If so, we'd probably need to explain a little more in that changeset.

Am waiting to look at the rest of the docs too closely while I see Matthew still asking questions!

**Comment by ematipico** [comment]
@sarah11918 I have updated the changeset regarding the bug fix.

**Comment by ematipico** [comment]
!preview

**Review by ematipico — COMMENTED** [review]
!preview

File: .changeset/blue-pens-divide.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+export default defineConfig({
+  base: '/blog'
+})
+```
+
+```diff
+// src/middleware.js
+export function onRequest(ctx, next) {
+-  return ctx.rerwrite("/about")
++  return ctx.rerwrite("/blog/about")
```
Comment (bluwy) [inline .changeset/blue-pens-divide.md:20]:
```suggestion
-  return ctx.rewrite("/about")
+  return ctx.rewrite("/blog/about")
```

File: packages/astro/test/fixtures/i18n-routing-fallback/astro.config.mjs [hunk 5]
```diff
@@ -8,8 +8,8 @@ export default defineConfig({
                 'en', 'pt', 'it'
             ],
             fallback: {
-                "it": "en",
-                pt: "en"
+                // "it": "en",
+                // pt: "en"
```
Comment (bluwy) [inline packages/astro/test/fixtures/i18n-routing-fallback/astro.config.mjs:12]: Is this commented out intentional?

File: packages/astro/src/core/routing/rewrite.ts [hunk 6]
```diff
@@ -12,44 +12,65 @@ export type FindRouteToRewrite = {
...
+	if (base !== '/' && newUrl.pathname.startsWith(base)) {
+		pathname = shouldAppendForwardSlash(trailingSlash, buildFormat)
+			? appendForwardSlash(newUrl.pathname)
+			: base !== '/'
+				? removeTrailingForwardSlash(newUrl.pathname)
+				: newUrl.pathname;
+		if (base !== '/') {
+			pathname = pathname.slice(base.length);
+		}
 	}
```
Comment (bluwy) [inline packages/astro/src/core/routing/rewrite.ts:52]: Since the outer block already checks `base !== '/'`, could we remove the inner ones?

**Review by bluwy — APPROVED** [review]

File: packages/astro/src/@types/astro.ts [hunk 7]
```diff
@@ -1665,6 +1665,20 @@ export interface AstroUserConfig {
 					 * */
 					redirectToDefaultLocale?: boolean;
 
+					/**
+					 * @docs
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1669]: I have some questions first here @ematipico 

Is this top-level i18n config? Is this entry in the correct place? Right now, it would be showing up between two `i18n.routing.*` entries.

Shouldn't it also be near `i18n.fallback`? And, this doesn't replace it, but is a separate, same-level configuration? (that's up around line 1567).

I think we should also see an example of both this and fallback configured together in one code sample, since I'm assuming they'd need to work together. Without any `i18n.fallback` configured, you default to showing 404s for pages that don't exist, so you can't have a strategy?

File: .changeset/lazy-feet-join.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+'astro': minor
+---
+
+Adds a new option to `i18n` called `routing.fallbackType` that allows to control the fallback logic. The new option accepts
+`"redirect"` or `"rewrite"`. The `"redirect"` option is the default value and matches the current behaviour of the fallback system.
+
+The option `"rewrite"` uses the new [rewriting system](https://docs.astro.build/en/guides/routing/#rewrites), and it generates
+the fallback pages based on the configuration.
+
+For example, the following configuration will generate a page `/fr/index.html` that will contain the same HTML rendered by the page `/en/index.html`
```
Comment (sarah11918) [inline .changeset/lazy-feet-join.md:11]:
```suggestion
For example, the following configuration will generate a page `/fr/index.html` that will contain the same HTML rendered by the page `/en/index.html` when `src/pages/fr/index.astro` does not exist.
```

**Review by sarah11918 — COMMENTED** [review]
This looks great @ematipico ! It's so exciting to see this feature continue to develop and allow people to control more and more!

I left some suggestions below for your consideration, so as always, correct what needs correcting and see what you think!

**Comment by sarah11918** [comment]
Should this be added to the 4.15 milestone? I have it in mine for docs!

File: .changeset/blue-pens-divide.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
+---
+'astro': patch
+---
+
+Fixes a bug in the rewrite logic, where Astro didn't correctly rewrite to website with `base`.
```
[the remark this answers is not in the document]
Comment (ematipico) [reply .changeset/blue-pens-divide.md:5]: @sarah11918 The fix is the other way around. The **fix** and correct behaviour is that `base` **must** be included in the URL.

```diff
- Astro.rewrite("/blog") // incorrect
+ Astro.rewrite("/corporate/blog") // correct
```

File: packages/astro/src/@types/astro.ts [hunk 10]
```diff
@@ -1665,6 +1665,42 @@ export interface AstroUserConfig {
...
+					 * @name i18n.routing.fallbackType
```
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:1670]: How do you know when to add a level or not?

File: .changeset/blue-pens-divide.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (sarah11918) [reply .changeset/blue-pens-divide.md:5]: Right, I think we're saying the same thing? the error was that base was included by Astro, so the user didn't have to write it. Now, after fixing, Astro no longer includes it, so the user must?

Or, are you saying if you had a base configured before, it was NEVER possible to get `Astro.rewrite` to work? That's why I had asked before, what if people *did* get it to work and now need to change because the "fix" breaks their project.

In other words, is it:

> Fixes a bug in the logic of `Astro.rewrite()` which did not work if you had a `base` configured.

> You can now use the `rewrite()` function on a project that has `base` configured. To do so, you must prepend the base to your existing rewrite URL:

??

[the remark this answers is not in the document]
Comment (sarah11918) [reply .changeset/blue-pens-divide.md:5]: > but when I read the suggestion, I understood that the **users** were led to add base in their URL.

Just checking, because in line 30 that is exactly what we're telling them to do, right? Users *do* need to ensure that base is in their URL?

[the remark this answers is not in the document]
Comment (sarah11918) [reply .changeset/blue-pens-divide.md:5]: @ematipico  I tried again! See if that is clearer!

[the remark this answers is not in the document]
Comment (ematipico) [reply .changeset/blue-pens-divide.md:5]: That's way better, thank you Sarah

File: packages/astro/src/@types/astro.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (sarah11918) [reply packages/astro/src/@types/astro.ts:1670]: @ematipico I know by checking the level around the others at the same level 😅 

Butt:
- Headers will show as `<h2>` (e.g. `i18`)
- Entries where you don't specify a header show as `<h3>` and will be listed in the page's table of contents (e.g. `i18n.routing`)
- If you need to go below that level, you'll need to specify exactly which heading you want, so `i18n.routing.fallbackType` needs `@kind h4` set manually
image

File: packages/astro/src/@types/astro.ts [hunk 11]
```diff
@@ -1665,6 +1665,43 @@ export interface AstroUserConfig {
...
+					 * @kind h4
+					 * @type {"redirect" | "rewrite"}
+					 * @default `"redirect"`
+					 * @version 4.15.0
+					 * @description
+					 *
+					 * When `i18n.fallback` is configured to avoid showing a 404 page for missing page routes, this option controls whether to [redirect](https://docs.astro.build/en/guides/routing/#redirects) to the fallback page, or to [rewrite](https://docs.astro.build/en/guides/routing/#rewrites) the fallback page's content in place.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:1677]:
```suggestion
					 * When [`i18n.fallback`](#i18nfallback) is configured to avoid showing a 404 page for missing page routes, this option controls whether to [redirect](https://docs.astro.build/en/guides/routing/#redirects) to the fallback page, or to [rewrite](https://docs.astro.build/en/guides/routing/#rewrites) the fallback page's content in place.
```
Since these two config options are a little bit far away from each other, I think a link here makes sense!

**Review by sarah11918 — COMMENTED** [review]
This looks great @ematipico !

The content here is now good, but I think we could also update `i18n.fallback` itself now to make sure that there's reference to this additional configuration someone might make besides just setting that. 

`i18n.fallback` is no longer just "the fallback strategy" because the strategy now combines these two (separated in diffrerent parts of the API) settings.

We could update the entry for `i18n.fallback` to be something like:

> An option to configure the fallback language content when navigating to pages that do not exist (e.g. a translated page has not been created).
>
> Use this object to declare a fallback locale route for each language you support. If no fallback is specified, then unavailable pages will return a 404.
>
> You can also optionally configure [`i18n.routing.fallbackType`](#i18nroutingfallbacktype), to select whether to execute this fallback content as a redirect (default) or a rewrite.

What do you think about something like this?

**Comment by ematipico** [comment]
> If I got to wave my magic wand (and, admitting I know nothing about what's underlying here), `i18n.fallback.type` (or `i18n.fallback.redirect` as a boolean, if you're not expecting any other option here) feels like a much more natural thing for users to configure.

I know, and I thought about it at the very beginning. However, `i18n.fallback` contains a list of locales and having a `type` doesn't fit in. That's why I changed it for `i18n.fallbackType`.

File: .changeset/blue-pens-divide.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
+---
+'astro': patch
+---
+
+Fixes a bug in the logic of `Astro.rewrite()` which led to the value for `base` , if configured, being automatically prepended to the rewrite URL passed. This was unintended behavior and has been corrected, and Astro now processes the URLs exactly as passed.
```
Comment (sarah11918) [inline .changeset/blue-pens-divide.md:5]:
```suggestion
Fixes a bug in the logic of `Astro.rewrite()` which led to the value for `base`, if configured, being automatically prepended to the rewrite URL passed. This was unintended behavior and has been corrected, and Astro now processes the URLs exactly as passed.
```

**Review by sarah11918 — APPROVED** [review]
Just an extra space, but approving for docs! Thank you @ematipico !
